blockchain/proxy.go

```diff
@@ -717,6 +743,203 @@ func (ep *proxy) IsDataTrieMigrated(ctx context.Context, address sdkCore.Address
+	}
+
+	for blockNum := fromBlock; blockNum <= toBlock; blockNum++ {
+		blockLogs, err := ep.getLogsFromBlock(ctx, filter.ShardID, blockNum, filter)
+		if err != nil {
+			return nil, err
+		}
+		matchingEvents = append(matchingEvents, blockLogs...)
+	}
+
+	return matchingEvents, nil
+}
+
+func resolveBlockRange(filter *sdkCore.FilterQuery, latestBlock uint64) (uint64, uint64, error) {
+	var genesisBlock uint64 = 0
+
+	if filter.ToBlock.HasValue && filter.ToBlock.Value > latestBlock {
+		return 0, 0, errors.New("toBlock is greater than the latest block")
+	}
+
+	// Check if both fromBlock and toBlock are set
+	if filter.FromBlock.HasValue && filter.ToBlock.HasValue {
+		if filter.ToBlock.Value-filter.FromBlock.Value <= MaximumBlocksDelta {
+			return filter.FromBlock.Value, filter.ToBlock.Value, nil
+		}
+		return 0, 0, errors.New("invalid block range: too many blocks to process")
+	}
+
+	// Check if only fromBlock is set
+	if filter.FromBlock.HasValue {
+		toBlock := latestBlock // Set toBlock to latestBlock
+		if toBlock-filter.FromBlock.Value <= MaximumBlocksDelta {
+			return filter.FromBlock.Value, toBlock, nil
+		}
+		return 0, 0, errors.New("invalid block range: too many blocks to process")
+	}
+
+	// Check if only toBlock is set
+	if filter.ToBlock.HasValue {
+		fromBlock := genesisBlock // Set fromBlock to genesisBlock
+		if filter.ToBlock.Value-fromBlock <= MaximumBlocksDelta {
+			return fromBlock, filter.ToBlock.Value, nil
+		}
+		return 0, 0, errors.New("invalid block range: too many blocks to process")
+	}
+
+	return 0, 0, errors.New("no block range specified")
+}
+
+// getBlockNumberByHash retrieves the block number associated with the given block hash
+func (ep *proxy) getBlockNumberByHash(ctx context.Context, shardID uint32, blockHash *[32]byte) (uint64, error) {
+	blockHashStr := hex.EncodeToString(blockHash[:])
+
+	buff, code, err := ep.GetHTTP(ctx, ep.endpointProvider.GetBlockByHash(shardID, blockHashStr))
+	if err != nil || code != http.StatusOK {
+		return 0, createHTTPStatusError(code, err)
+	}
+
+	var response data.BlockResponse
+	if err := json.Unmarshal(buff, &response); err != nil {
+		return 0, err
+	}
+
+	blockNonce := response.Data.Block.Nonce
+
+	// Cache the raw response bytes
+	if ep.filterQueryBlockCacher != nil && len(buff) > 0 {
+		cacheKey := make([]byte, 8)
+		binary.BigEndian.PutUint64(cacheKey, blockNonce)
+		ep.filterQueryBlockCacher.Put(cacheKey, buff, len(buff))
+	}
+
+	return blockNonce, nil
+}
+
+// getLogsFromBlock retrieves logs from a specific block and filters them
+func (ep *proxy) getLogsFromBlock(ctx context.Context, shardID uint32, blockNum uint64, filter *sdkCore.FilterQuery) ([]*transaction.Events, error) {
+	cacheKey := make([]byte, 8)
+	binary.BigEndian.PutUint64(cacheKey, blockNum)
+
+	if ep.filterQueryBlockCacher != nil {
+		cachedResponse, found := ep.filterQueryBlockCacher.Get(cacheKey)
+		if found {
+			cachedBuff, ok := cachedResponse.([]byte)
```

Comment:
duplicated code logic in this function. I would make 3 steps in this function:
1. get the block response bytes (by either getting them from the cacher or by querying the API endpoint and then save it in the cacher) -> I would extract this in a new method
2. unmarshal the block 
3. call the method to extract the logs